ts/Core/Chart/Chart.ts

```diff
@@ -825,66 +819,10 @@ class Chart {
             chart.temporaryDisplay();
         }
 
-        // Adjust title layout (reflow multiline text)
-        chart.layOutTitles();
-
-        // link stacked series
-        i = series.length;
-        while (i--) {
-            serie = series[i];
-
-            if (serie.options.stacking || serie.options.centerInCategory) {
-                hasStackedSeries = true;
-
-                if (serie.isDirty) {
-                    hasDirtyStacks = true;
-                    break;
-                }
-            }
-        }
-        if (hasDirtyStacks) { // mark others as dirty
-            i = series.length;
-            while (i--) {
-                serie = series[i];
-                if (serie.options.stacking) {
-                    serie.isDirty = true;
-                }
-            }
-        }
-
-        // Handle updated data in the series
-        series.forEach(function (serie): void {
-            if (serie.isDirty) {
-                if (serie.options.legendType === 'point') {
-                    if (typeof serie.updateTotals === 'function') {
-                        serie.updateTotals();
-                    }
-                    redrawLegend = true;
-                } else if (
-                    legendUserOptions &&
-                    (
-                        legendUserOptions.labelFormatter ||
-                        legendUserOptions.labelFormat
-                    )
-                ) {
-                    redrawLegend = true; // #2165
-                }
-            }
-            if (serie.isDirtyData) {
-                fireEvent(serie, 'updatedData');
-            }
-        });
-
-        // handle added or removed series
-        if (redrawLegend && legend && legend.options.enabled) {
-            // draw legend graphics
-            legend.render();
-
-            chart.isDirtyLegend = false;
-        }
+        chart.setOffsets();
```

Comment:
I think all the changes to Chart.ts are redundant. Unless I am missing something, the whole point of this refactoring is to add the `setOffsets` event. But the handling itself is flawed, it is not enough to set `yAxis.options.offsets`: https://jsfiddle.net/highcharts/LLExL/14497/ .

Instead, I think we must do like the range selector did, to use the `Chart` `getMargins` event and add to `this.plotTop`.

Reply:
I remember discussing this issue with Karol and Grzesiek, they had similar problems as Sebastian had with StockTools and RangeSelector. `getMargins()` is not sufficient in some cases, and we end up with multiple events added to fix minor issues.

- In StockTools we have callbacks dedicated to: `getMargins`, `beforeRender` and `beforeRedraw` (last two added this year: https://github.com/highcharts/highcharts/pull/14920) events
- In RangeSelector, we have a small monster, called `alignElements()`, where we check other components (Captions, Legend, Exporting button), but I guess this might inevitable (collision detection). And of course, we have events added to `beforeRender` and `getMargins` (I'm curious if adding `beforeRedraw` might resolve some issues we are not aware of and/or let us slim down the RangeSelector module)
- In Navigator, we have `getMargins` and `afterSetChartSize`

Instead, I suggested refactoring that part of the `chart.redraw()` and let all external components (Legend, Title/Subtitle, RangeSelector, Navigator, StockTools, Breadcrumbs) register the space they need. The order of the events would define the order of components that reserve space. The Legend, Title/Subtitle are Core components, so they are simply added in the correct place. I might be wrong, but to me, it seems other components just find workarounds on how to solve it. Perhaps, there's a better way to solve it? 🤔

Reply:
It would be great if we find a common API for that. A thing from my side, just lookin into the responsive module: It might be helpful if every component also takes optional a maximum to be available for them. This would allow us to cover narrowed layout situations and responsiveness.

Reply:
I have also been thinking about this. A possible solution is to introduce a Regions concept. The regions may be called something like north, northeast, east, southeast... (you get the picture :). Every element would get assigned to a region. So the default title, and the subtitle, would be region "north". The range selector would be the same. Legend `align: right` would internally mean region east. Inside each region, the elements would have an order. This would allow us to do standardized layout calculation of all elements in the various regions first (at the point in time when we currently run `Chart.getMargins`), then find out how much is left for the plot area, which is the center region.

Reply:
@pawelfus I see you are suggesting something further down here, but I can't comment there...

Reply:
GitHub is duplicating my comment, I don't know why... It's the same as the one above. Regions idea looks good, do you have some API in your mind?

Reply:
I agree with that point. I will prepare it similar to the rangeSelector for now, until we will decide to implement `regions` approach.